src/test/java/com/commercetools/sync/services/impl/BaseServiceImplTest.java

```diff
@@ -0,0 +1,302 @@
+        );
+        verify(client).execute(any(TestQuery.class));
+    }
+
+    @Test
+    void fetchResource_WithNullKey_ShouldFetchNothing() {
+        Optional<TestResource> optional = service.fetchResource(null,
+            () -> null, TestResource::getKey).toCompletableFuture().join();
+        assertThat(optional).isEmpty();
+    }
+
+    @Test
+    void fetchResource_WithKey_ShouldFetchResource() {
+        String resourceId = RandomStringUtils.random(15);
+        String resourceKey = RandomStringUtils.random(15);
+
+        TestResource mock = mock(TestResource.class);
+        when(mock.getId()).thenReturn(resourceId);
+        when(mock.getKey()).thenReturn(resourceKey);
+        TestPagedQueryResult result = mock(TestPagedQueryResult.class);
+        when(result.head()).thenReturn(Optional.of(mock));
+
+        when(client.execute(any())).thenReturn(completedFuture(result));
+
+        Optional<TestResource> resourceOptional = service.fetchResource(resourceKey,
+            () -> mock(TestQuery.class), TestResource::getKey).toCompletableFuture().join();
+
+        assertAll(
+            () -> assertThat(resourceOptional).isNotEmpty(),
+            () -> assertThat(resourceOptional).containsSame(mock),
+            () -> assertThat(service.keyToIdCache.get(resourceKey)).isEqualTo(resourceId)
+        );
+        verify(client).execute(any(TestQuery.class));
+    }
+
+    @Test
+    void buildResourceKeysQueryPredicate_WithEmptyKeySet_ShouldBuildQueryPredicate() {
+        QueryPredicate<TestResource> queryPredicate = service.buildResourceKeysQueryPredicate(new HashSet<>());
+
+        assertThat(queryPredicate.toSphereQuery()).isEqualTo("key in ()");
+    }
+
+    @Test
+    void buildResourceKeysQueryPredicate_WithKeySet_ShouldBuildQueryPredicate() {
+        String key1 = RandomStringUtils.random(15);
+        String key2 = RandomStringUtils.random(15);
+
+        HashSet<String> stateKeys = new HashSet<>();
+        stateKeys.add(key1);
+        stateKeys.add(key2);
+
+        QueryPredicate<TestResource> queryPredicate = service.buildResourceKeysQueryPredicate(stateKeys);
+
+        assertThat(queryPredicate.toSphereQuery())
+            .isIn("key in (\"" + key1 + "\", \"" + key2 + "\")", "key in (\"" + key2 + "\", \"" + key1 + "\")");
+    }
+
+    @Test
+    void buildResourceKeysQueryPredicate_WithKeySetContainingInvalidKeys_ShouldBuildQueryPredicate() {
+        String key1 = RandomStringUtils.random(15);
+        String key2 = RandomStringUtils.random(15);
+
+        HashSet<String> stateKeys = new HashSet<>();
+        stateKeys.add(key1);
+        stateKeys.add(key2);
+        stateKeys.add("");
+        stateKeys.add(null);
+
+        QueryPredicate<TestResource> queryPredicate = service.buildResourceKeysQueryPredicate(stateKeys);
+
+        assertThat(queryPredicate.toSphereQuery())
+            .isIn("key in (\"" + key1 + "\", \"" + key2 + "\")", "key in (\"" + key2 + "\", \"" + key1 + "\")");
+    }
+
```

Comment:
Yes, because `#fetchCachedResourceId` is internally calling this method, so tests named `fetchCachedResourceId_WithFetchReturnResource*` are testing `#fetchAndCache` (there are no test methods called `fetchAndCache_*`).
They can be renamed and changed to directly call `#fetchAndCache` but than `#fetchCachedResourceId` wouldn't be fully tested (not all cases covered) or we could add a test for not covered case but it would be a duplication of other test.